plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/PassthroughQueryPageSource.java

```
@@ -87,5 +114,121 @@ public long getSystemMemoryUsage()
+    {
+        private final ElasticsearchClient client;
+        private final String first;
+
+        private String scrollId;
+        private Boolean hasMoreHits;
+        private long readTimeNanos;
+        String response;
+        ObjectMapper objectMapper;
+
+        public SearchHitIterator(ElasticsearchClient client, String first)
+        {
+            this.client = client;
+            this.first = first;
+            this.objectMapper = new ObjectMapper();
+        }
+
+        public long getReadTimeNanos()
+        {
+            return readTimeNanos;
+        }
+
+        @Override
+        protected String computeNext()
+        {
+            long start = System.nanoTime();
+
+            if (scrollId == null) {
+                response = first;
+            }
+            else {
+                response = client.executeScrollQuery(scrollId);
+            }
+
+            readTimeNanos += System.nanoTime() - start;
+            reset(response);
+
+            if (!hasMoreHits || scrollId == null) {
+                return endOfData();
+            }
+
+            return response;
+        }
+
+        private void reset(String response)
+        {
+            ScrollSearchResponse scrollSearchResponse = extractSearchResponse(response);
+            if (scrollSearchResponse != null) {
+                scrollId = scrollSearchResponse.getScrollId();
+                hasMoreHits = scrollSearchResponse.hasMoreHits();
+            }
+        }
+
+        private ScrollSearchResponse extractSearchResponse(String response)
+        {
+            try {
+                ScrollSearchResponse scrollSearchResponse = objectMapper.readValue(response, ScrollSearchResponse.class);
+                if (scrollSearchResponse != null) {
+                    return scrollSearchResponse;
+                }
+            }
+            catch (JsonProcessingException e) {
+                LOG.warn(e, "No Scroll_id returned from elasticSearch");
```

better to throw a trinoexception here instead of returning incomplete results